modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+            }
+        };
+
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl TryFrom<RawDenomTrace> for DenomTrace {
+    type Error = Error;
+
+    fn try_from(value: RawDenomTrace) -> Result<Self, Self::Error> {
+        let base_denom = Denom::from_str(&value.base_denom)?;
+        let trace_path = TracePath::from_str(&value.path)?;
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl From<DenomTrace> for RawDenomTrace {
+    fn from(value: DenomTrace) -> Self {
+        Self {
+            path: value.trace_path.to_string(),
+            base_denom: value.base_denom.to_string(),
+        }
+    }
+}
+
+impl From<Denom> for DenomTrace {
+    fn from(denom: Denom) -> Self {
+        Self {
+            trace_path: Default::default(),
+            base_denom: denom,
+        }
+    }
+}
+
+impl fmt::Display for DenomTrace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.trace_path.0.is_empty() {
+            write!(f, "{}", self.base_denom)
+        } else {
+            write!(f, "{}/{}", self.trace_path, self.base_denom)
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct HashedDenom(Vec<u8>);
+
+impl From<&DenomTrace> for HashedDenom {
+    fn from(value: &DenomTrace) -> Self {
+        let mut hasher = Sha256::new();
+        hasher.update(value.to_string().as_bytes());
+        let denom_bytes = hasher.finalize();
+        Self(denom_bytes.to_vec())
+    }
+}
+
+impl fmt::Display for HashedDenom {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let denom_hex = String::from_utf8(hex::encode_upper(&self.0)).map_err(|_| fmt::Error)?;
+        write!(f, "{}/{}", IBC_DENOM_PREFIX, denom_hex)
+    }
+}
+
+impl FromStr for HashedDenom {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<&str> = s.split('/').collect();
+        if parts.len() != 2 || parts[1].trim().is_empty() {
+            return Err(Error::malformed_hash_denom());
+        } else if parts[0] != IBC_DENOM_PREFIX {
+            return Err(Error::missing_denom_ibc_prefix());
+        }
+
+        let bytes = hex::decode_upper(parts[1]).map_err(Error::parse_hex)?;
+        Ok(Self(bytes))
+    }
+}
+
+/// A decimal type for representing token transfer amounts.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Display, From)]
+pub struct Amount(U256);
+
+impl FromStr for Amount {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let amount = U256::from_str_radix(s, 10).map_err(Error::invalid_amount)?;
+        Ok(Self(amount))
+    }
+}
+
+impl From<u64> for Amount {
+    fn from(v: u64) -> Self {
+        Self(v.into())
+    }
+}
+
+// We only provide an `Add<Decimal>` implementation which always panics on overflow.
+impl Add<Self> for Amount {
```

Comment:
Seems like a vector of attack to crash a chain. I don't think a library should ever crash on error. We could instead expose `checked_add()` and more, and let the chain decide how it wants to handle overflows.